File: SQL Server 2012/Projects/Visual Basic/Chapter 22/ApClrObjects/Stored Procedures/GetInvoiceTotalSum.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server


Partial Public Class StoredProcedures
    <Microsoft.SqlServer.Server.SqlProcedure()> _
    Public Shared Sub GetInvoiceTotalSum(ByRef invoiceTotalSum As SqlMoney,
            ByVal startDate As SqlDateTime, ByVal vendorName As SqlString)
        ' Open the Connection object for the context
        Dim connection As New SqlConnection("Context connection=true")
        connection.Open()

        ' Create a string that defines the Select statement
        Dim sql As String =
            "SELECT SUM(InvoiceTotal) " &
            "FROM Invoices Join Vendors " &
            "  ON Invoices.VendorID = Vendors.VendorID " &
            "WHERE ((InvoiceDate >= @StartDate) AND " &
            "       (VendorName LIKE @VendorName))"

        ' Create the Command object
        Dim command As New SqlCommand(sql, connection)

        ' Add new parameters to the Parameters collection
        Dim param As New SqlParameter("@StartDate", SqlDbType.DateTime)
        param.Value = startDate
        command.Parameters.Add(param)

        param = New SqlParameter("@VendorName", SqlDbType.VarChar)
        param.Value = vendorName
        command.Parameters.Add(param)

        ' Assign a value to the output parameter
        Try
            Dim scalarValue As Decimal = CDec(command.ExecuteScalar)
            invoiceTotalSum = CType(scalarValue, SqlMoney)
        Catch ex As Exception    ' If a null value causes the cast to throw an exception
            invoiceTotalSum = 0
        End Try

        ' Close the Connection object
        connection.Close()
    End Sub
End Class

File: SQL Server 2012/Projects/C#/Chapter 21/ApClrObjects/ApClrObjects/GetInvoiceTotalSum.cs
//------------------------------------------------------------------------------
// <copyright file="CSSqlStoredProcedure.cs" company="Microsoft">
//     Copyright (c) Microsoft Corporation.  All rights reserved.
// </copyright>
//------------------------------------------------------------------------------
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

public partial class StoredProcedures
{
    [Microsoft.SqlServer.Server.SqlProcedure]
    public static void GetInvoiceTotalSum(
        out SqlMoney invoiceTotalSum,
        SqlDateTime startDate,
        SqlString vendorName)
    {
        // open the Connection object for the context
        SqlConnection connection = new SqlConnection(
            "Context connection=true");
        connection.Open();

        // create the Command object
        SqlCommand command = new SqlCommand();
        command.Connection = connection;

        // create a string that defines the SELECT statement
        string sql =
            "SELECT SUM(InvoiceTotal) " +
            "FROM Invoices JOIN Vendors " +
            "    ON Invoices.VendorID = Vendors.VendorID " +
            "WHERE ((InvoiceDate >= @StartDate) AND " +
            "       (VendorName LIKE @VendorName))";

        // add the query to the Command object
        command.CommandText = sql;

        // add new parameters to the Parameters collection
        SqlParameter param = new SqlParameter(
            "@StartDate", SqlDbType.DateTime);
        param.Value = startDate;
        command.Parameters.Add(param);

        param = new SqlParameter(
            "@VendorName", SqlDbType.VarChar);
        param.Value = vendorName;
        command.Parameters.Add(param);

        // assign a value to the output parameter
        try
        {
            decimal scalarValue = (decimal)command.ExecuteScalar();
            invoiceTotalSum = (SqlMoney)scalarValue;
        }
        catch  // if a null value is returned
        {
            invoiceTotalSum = 0;
        }

        // close the Connection object
        connection.Close();
    }
}
